sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java

```diff
@@ -0,0 +1,681 @@
+                List<HRegionLocation> regionLocations = getRegionLocations();
+
+                // builds set of regions who are part of the table scan
+                Set<byte[]> tableRegions = new TreeSet<>(Bytes.BYTES_COMPARATOR);
+                for (HRegionLocation regionLocation : regionLocations) {
+                    tableRegions.add(regionLocation.getRegionInfo().getRegionName());
+                }
+
+                // calculate estimated size for the regions
+                Admin admin = connection.getAdmin();
+                ClusterStatus clusterStatus = admin.getClusterStatus();
+                Collection<ServerName> servers = clusterStatus.getServers();
+                for (ServerName serverName : servers) {
+                    ServerLoad serverLoad = clusterStatus.getLoad(serverName);
+                    for (RegionLoad regionLoad : serverLoad.getRegionsLoad().values()) {
+                        byte[] regionId = regionLoad.getName();
+                        if (tableRegions.contains(regionId)) {
+                            long regionSizeBytes = regionLoad.getStorefileSizeMB() * 1_048_576L;
+                            estimatedSizeBytes += regionSizeBytes;
+                        }
+                    }
+                }
+            }
+            return estimatedSizeBytes;
+        }
+
+        private List<HRegionLocation> getRegionLocations()
+                throws Exception {
+            final Scan scan = read.serializableScan.getScan();
+            byte[] startRow = scan.getStartRow();
+            byte[] stopRow = scan.getStopRow();
+
+            final List<HRegionLocation> regionLocations = new ArrayList<>();
+
+            final boolean scanWithNoLowerBound = startRow.length == 0;
+            final boolean scanWithNoUpperBound = stopRow.length == 0;
+            Configuration configuration = this.read.serializableConfiguration.getConfiguration();
+            try (Connection connection = ConnectionFactory.createConnection(configuration)) {
+                TableName tableName = TableName.valueOf(read.tableId);
+                RegionLocator regionLocator = connection.getRegionLocator(tableName);
+                List<HRegionLocation> tableRegionInfos = regionLocator.getAllRegionLocations();
+                for (HRegionLocation regionLocation : tableRegionInfos) {
+                    final byte[] startKey = regionLocation.getRegionInfo().getStartKey();
+                    final byte[] endKey = regionLocation.getRegionInfo().getEndKey();
+                    boolean isLastRegion = endKey.length == 0;
+                    // filters regions who are part of the scan
+                    if ((scanWithNoLowerBound
+                            || isLastRegion || Bytes.compareTo(startRow, endKey) < 0)
+                            && (scanWithNoUpperBound || Bytes.compareTo(stopRow, startKey) > 0)) {
+                        regionLocations.add(regionLocation);
```

Comment:
At the begining I didn't want to use both ByteKey and ByteKeyRange because I thought they were related to Bigtable not Beam.  I propose you to change this logic afterwards, in a subsequent PR, maybe with the subsplits work, I don't really want to change the current logic.